woocommerce/woocommerce-ios
# Product description AI: feature integration

Part of #9465 


Apology for the 500+ diffs, the unit tests were a bit long with coverage on a few different properties.

## Why


After previous PRs for https://github.com/woocommerce/woocommerce-ios/issues/9465, this PR integrated the bottom sheet with the product description AI flow.

## The flow

There are two input fields:

- Product name
- Product features with default value as the pre-existing product description

There is only 1 CTA in the beginning to generate a product description, when the product name & features are non-empty.

After a description is generated, there are 3 CTAs:
- Copy generated description
- Toggle description generation (start/stop)
- Apply the generated description and the latest name to the product

## How

`ProductDescriptionGenerationView` is the main SwiftUI view, with a view model `ProductDescriptionGenerationViewModel` to supply observable data and handle user interactions. To integrate the view with the bottom sheet, `ProductDescriptionGenerationHostingController ` is shown for the CTA in the bottom sheet in `AztecEditorViewController`. Since the product name can also be updated to polish the description, the product name is also added to `Editor.OnContentSave`. When applying the content to the product, `ProductFormViewController` updated its name and description accordingly.

## Testing instructions


- Log in to a WPCOM store if needed
- Go to the Products tab
- Create a product if needed
- Tap on an editable product
- Tap on the product description row
- Tap on the magic wand CTA above the keyboard --> a bottom sheet should be shown, the product name and features should be pre-filled with the pre-existing name and description
- Enter or update the product name
- Enter or update the product features
- Tap `Generate` --> a spinner should be shown in the button, and after a while (30-60 seconds from my testing), some generated text should be shown below in gray with 3 CTAs: `Copy`, retry icon, and `Apply`
- Tap `Copy` and try pasting the content somewhere in or outside of the app --> the generated text should be in the pasteboard
- Tap the retry icon in the middle --> the icon should become a pause button. after a while, the generated text should be replaced above
- Tap the retry icon again, and immediately tap again  --> the icon should go back to the retry icon
- Tap `Apply` --> the product description should be replaced with the generated text
- Tap `Done` --> the product form should have the updated product name and generated text as its description, with the `Save` CTA enabled
- Tap `Save` --> the product name & description should be saved remotely

## Screenshots



https://user-images.githubusercontent.com/1945542/234768978-3dc53b86-92b2-49ec-a567-ce6f50f94eb4.mov


Initial state | After generation
-- | --
![Simulator Screen Shot - iPhone 14 - 2023-04-27 at 13 13 50](https://user-images.githubusercontent.com/1945542/234768951-3af58e7d-a46e-4aaf-bc4d-c616bb6d8f44.png) | ![Simulator Screen Shot - iPhone 14 - 2023-04-27 at 13 13 54](https://user-images.githubusercontent.com/1945542/234768956-417c8093-29ff-4e79-b25f-0da68ccc7f66.png)



---
